oracle/oracles/pd.go

```diff
@@ -47,19 +48,111 @@ import (
+	// that is new enough for checking the legality of the read ts, causing them have to fetch the latest ts from PD,
+	// which is time-consuming.
+	adaptiveUpdateTSIntervalStateAdapting
+	// adaptiveUpdateTSIntervalStateRecovering represents the state that the update ts interval have once been shrunk,
+	// to adapt to reads with short staleness, but there isn't any such read operations for a while, so that we
+	// gradually recover the update interval to the configured value.
+	adaptiveUpdateTSIntervalStateRecovering
+	// adaptiveUpdateTSIntervalStateUnadjustable represents the state that the user has configured a very short update
+	// interval, so that we don't have any space to automatically adjust it.
+	adaptiveUpdateTSIntervalStateUnadjustable
+)
+
+func (s adaptiveUpdateTSIntervalState) String() string {
+	switch s {
+	case adaptiveUpdateTSIntervalStateNormal:
+		return "normal"
+	case adaptiveUpdateTSIntervalStateAdapting:
+		return "adapting"
+	case adaptiveUpdateTSIntervalStateRecovering:
+		return "recovering"
+	case adaptiveUpdateTSIntervalStateUnadjustable:
+		return "unadjustable"
+	default:
+		return fmt.Sprintf("unknown(%v)", int(s))
+	}
+}
+
+const (
+	// minAllowedAdaptiveUpdateTSInterval is the lower bound of the adaptive update ts interval for avoiding an abnormal
+	// read operation causing the update interval to be too short.
+	minAllowedAdaptiveUpdateTSInterval = 500 * time.Millisecond
+	// adaptiveUpdateTSIntervalShrinkingPreserve is the duration that we additionally shrinks when adapting to a read
+	// operation that requires a short staleness.
+	adaptiveUpdateTSIntervalShrinkingPreserve = 100 * time.Millisecond
+	// adaptiveUpdateTSIntervalBlockRecoverThreshold is the threshold of the difference between the current update
+	// interval and the staleness the read operation request to prevent the update interval from recovering back to
+	// normal.
+	adaptiveUpdateTSIntervalBlockRecoverThreshold = 200 * time.Millisecond
+	// adaptiveUpdateTSIntervalRecoverPerSecond is the duration that the update interval should grow per second when
+	// recovering to normal state from adapting state.
+	adaptiveUpdateTSIntervalRecoverPerSecond = 20 * time.Millisecond
+	// adaptiveUpdateTSIntervalDelayBeforeRecovering is the duration that we should hold the current adaptive update
+	// interval before turning back to normal state.
+	adaptiveUpdateTSIntervalDelayBeforeRecovering = 5 * time.Minute
+)
+
 // pdOracle is an Oracle that uses a placement driver client as source.
 type pdOracle struct {
 	c pd.Client
 	// txn_scope (string) -> lastTSPointer (*lastTSOPointer)
 	lastTSMap sync.Map
 	quit      chan struct{}
+	// The configured interval to update the low resolution ts. Set by SetLowResolutionTimestampUpdateInterval.
+	// For TiDB, this is directly controlled by the system variable `tidb_low_resolution_tso_update_interval`.
```

Comment:
You can consider picking from other cherrypicks to old branches, in which I've modified some contents. e.g. https://github.com/tikv/client-go/pull/1491

Reply:
That makes sense. I've picked #1491 as the last commit.